vic/drivers/shared_all/src/calc_root_fraction.c

```diff
@@ -41,128 +41,126 @@ calc_root_fractions(veg_con_struct  *veg_con,
-                /** CASE 1: Root Zone Completely in Soil Layer **/
-                sum_fract += veg_con[veg].zone_fract[zone];
+            // Determine the depth interval for consideration
+            // Depth intervals span from the previous layer or zone boundary
+            // to the next layer or zone boundary.
+            if (Lsum <= Zsum && layer < options.Nlayer) {
+                Dsum = Lsum;
             }
             else {
-                /** CASE 2: Root Zone Partially in Soil Layer **/
-                if (Zsum < Lsum - Lstep) {
-                    /** Root zone starts in previous soil layer **/
-                    Zmin_depth = Lsum - Lstep;
-                    Zmin_fract = linear_interp(Zmin_depth, Zsum, Zsum + Zstep,
-                                               0,
-                                               veg_con[veg].zone_fract[zone]);
-                }
-                else {
-                    /** Root zone starts in current soil layer **/
-                    Zmin_depth = Zsum;
-                    Zmin_fract = 0.;
-                }
-                if (Zsum + Zstep <= Lsum) {
-                    /** Root zone ends in current layer **/
-                    Zmax = Zsum + Zstep;
-                }
-                else {
-                    /** Root zone extends beyond bottom of current layer **/
-                    Zmax = Lsum;
-                }
-                sum_fract += linear_interp(Zmax, Zsum, Zsum + Zstep, 0,
-                                           veg_con[veg].zone_fract[zone]) -
-                             Zmin_fract;
+                Dsum = Zsum;
             }
+            dD = Dsum - Dsumprev;
+            Dsumprev = Dsum;
 
-            /** Update Root Zone and Soil Layer **/
-            if (Zsum + Zstep < Lsum) {
-                Zsum += Zstep;
-                zone++;
+            // Add root fraction in this interval to the running total
+            // for this layer
+            if (zone < options.ROOT_ZONES) {
+                sum_fract += dD * root_dens[zone];
             }
-            else if (Zsum + Zstep == Lsum) {
-                Zsum += Zstep;
-                zone++;
-                if (layer < options.Nlayer) {
-                    veg_con[veg].root[layer] = sum_fract;
-                    sum_fract = 0.;
+
+            // Assign the total root fraction for this soil layer
+            // Wait to do this until either we've completed a soil layer
+            // (Lsum > Lsumprev && Dsum == Lsum)
+            // or we're at the final layer and we've completed all root zones
+            // (layer >= options.Nlayer - 1 &&
+            //  zone >= options.ROOT_ZONES - 1 && Dsum == Zsum)
+            if ( Lsum > Lsumprev && Dsum == Lsum &&
+                 ( layer < options.Nlayer - 1 ||
+                   ( layer >= options.Nlayer - 1 &&
+                     zone >= options.ROOT_ZONES - 1 &&
+                     Dsum == Zsum ) ) ) {
```

Comment:
There seems to be a parenthesis problem - should it be: `( Lsum > Lsumprev && Dsum == Lsum && ( layer < options.Nlayer - 1 ) || ( layer >= options.Nlayer - 1 && zone >= options.ROOT_ZONES - 1 && Dsum == Zsum ) ) `? (note the parenthesis after `layer < options.Nlayer -1`) Double check

Reply:
No, the parentheses are the way I intended them...  But I suppose I could remove the `layer >= options.Nlayer -1 condition in the second set of parentheses, since it's redundant...